File: src/Postprocess/shadowmap.cpp
//
// Created by dydxh on 12/4/18.
//

#include "shadowmap.h"
#include "../config.h"

Shadowmap::Shadowmap() {
    this->Initialize();
}
Shadowmap::~Shadowmap() {
    glDeleteFramebuffers(1, &DepthMapFBO);
}
void Shadowmap::Initialize() {
    glGenFramebuffers(1, &DepthMapFBO);
    DepthMap.GenerateShadowMap(SHADOW_WIDTH, SHADOW_HEIGHT);
    glBindFramebuffer(GL_FRAMEBUFFER, DepthMapFBO);
    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, DepthMap.ID, 0);
    glDrawBuffer(GL_NONE);
    glReadBuffer(GL_NONE);
    glBindFramebuffer(GL_FRAMEBUFFER, 0);
}

void Shadowmap::UpdateFrustum(const Scene *scene) {
    glm::mat4 lightviewmat = glm::lookAt(PARLIGHT_POSITION + scene->offset,
                                         PARLIGHT_POSITION + scene->offset + PARLIGHT_DIR,
                                         glm::vec3(0.0f, 1.0f, 0.0f));
    this->near = this->left  = this->bottom = 2000000.0f;
    this->far  = this->right = this->top    = -2000000.0f;
    for(int i = 0; i < CHUNK_SIZE; i++){
        for(int j = 0; j < CHUNK_SIZE; j++){
            for(int k = 0; k <= MESH_SIZE; k++){
                for(int h = 0; h <= MESH_SIZE; h++){
                    glm::vec3 pos = scene->offset +
                    scene->chunk_offset[i][j] +
                    scene->mesh_offset[k][h];
                    pos.y = scene->chunk[i][j]->height[k][h];
                    pos = glm::vec3(lightviewmat * glm::vec4(pos, 1.0f));
                    this->left   = glm::min(this->left,   pos.x);
                    this->right  = glm::max(this->right,  pos.x);
                    this->bottom = glm::min(this->bottom, pos.y);
                    this->top    = glm::max(this->top,    pos.y);
                    this->near   = glm::min(this->near,  -pos.z);
                    this->far    = glm::max(this->far ,  -pos.z);
                }
            }
        }
    }
    this->left -= 0.2f, this->bottom -= 0.2f, this->near -= 0.2f;
    this->right += 0.2f, this->top += 0.2f, this->far += 0.2f;
}

glm::mat4 Shadowmap::GetlightSpaceMatrix(const Scene* scene) {
    glm::mat4 lightProjection = glm::ortho(this->left, this->right,
               this->bottom, this->top,
               this->near, this->far);
    glm::mat4 lightviewmat = glm::lookAt(PARLIGHT_POSITION + scene->offset,
                                         PARLIGHT_POSITION + scene->offset + PARLIGHT_DIR,
                                         glm::vec3(0.0f, 1.0f, 0.0f));
    glm::mat4 lightSpaceMatrix = lightProjection * lightviewmat;
    return lightSpaceMatrix;
}

void Shadowmap::BeginMakeMap(){
    glViewport(0, 0, SHADOW_WIDTH, SHADOW_HEIGHT);
    glBindFramebuffer(GL_FRAMEBUFFER, DepthMapFBO);
    glClear(GL_DEPTH_BUFFER_BIT);
    glCullFace(GL_BACK);
}

void Shadowmap::EndMakeMap() {
//    glCullFace(GL_BACK);
    glBindFramebuffer(GL_FRAMEBUFFER, 0);
    glViewport(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
}

File: iOS/Polyworld/Shadowmap.swift
//
//  Shadowmap.swift
//  Polyworld
//
//  Created by 沈心逸 on 2019/1/8.
//  Copyright © 2019年 Sxy. All rights reserved.
//

import Foundation
import Metal
import simd

class Shadowmap {
    var depthMap, bluredShadow: MTLTexture!
    var near, far, left, right, top, bottom: Float!
    
    init() {
        depthMap = ResourceManager.device.makeTexture(descriptor: ResourceManager.depthBufferDescriptor)!
    }
    
    func UpdateFrustum(scene: Scene){
        let lightviewmat = lookAt(PARLIGHT_POSITION + scene.offset, PARLIGHT_POSITION + scene.offset + PARLIGHT_DIR, float3(0.0, 1.0, 0.0))
        near = 200000
        left = 200000
        bottom = 200000
        far = -200000
        right = -200000
        top = -200000
        for i in 0 ..< CHUNK_SIZE {
            for j in 0 ..< CHUNK_SIZE {
                for k in 0 ... MESH_SIZE {
                    for h in 0 ... MESH_SIZE {
                        var pos = scene.offset + scene.chunk_offset[i][j] + scene.mesh_offset[k][h]
                        pos.y = scene.chunk[i][j].height[k][h]
                        let tmpPos = pos
                        pos = (lightviewmat * float4(tmpPos.x, tmpPos.y, tmpPos.z, 1.0)).xyz
                        left   = min(left,   pos.x)
                        right  = max(right,  pos.x)
                        bottom = min(bottom, pos.y)
                        top    = max(top,    pos.y)
                        near   = min(near,  -pos.z)
                        far    = max(far ,  -pos.z)
                    }
                }
            }
        }
        near -= 0.2
        left -= 0.2
        bottom -= 0.2
        far += 0.2
        right += 0.2
        top += 0.2
    }
    
    func GetlightSpaceMatrix(scene: Scene) -> float4x4 {
        let lightProjection = float4x4(orthoProjectionLeft: left, right: right, bottom:     bottom, top: top, near: near, far: far)
        let lightviewmat = lookAt(PARLIGHT_POSITION + scene.offset, PARLIGHT_POSITION + scene.offset + PARLIGHT_DIR, float3(0.0, 1.0, 0.0))
        let lightSpaceMatrix = lightProjection * lightviewmat
        return lightSpaceMatrix
    }
    
    func BeginMakeMap() {
        
    }
    
    func EndMakeMap() {
        
    }
}
